sources/packages/backend/apps/queue-consumers/src/services/application/application.service.ts

```diff
@@ -103,13 +106,40 @@ export class ApplicationService {
             submittedStatus: StudentAssessmentStatus.Submitted,
           },
         )
-        .andWhere(`NOT EXISTS (${inProgressStatusesExistsQuery})`)
+        .andWhere(`NOT EXISTS (${this.inProgressStatusesExistsQuery})`)
         .setParameter("inProgressStatuses", [
           StudentAssessmentStatus.Queued,
           StudentAssessmentStatus.InProgress,
+          StudentAssessmentStatus.CancellationRequested,
+          StudentAssessmentStatus.CancellationQueued,
         ])
         .orderBy("studentAssessment.createdAt")
         .getMany()
     );
   }
+
+  /**
+   * Finds all applications with some student assessment with cancellation requested.
+   * All student assessments to be cancelled will be returned ordered by its creation
+   * date to allow the select of the next one to be cancelled.
+   * @returns applications with assessment to be cancelled.
+   */
+  async getApplicationsToCancelAssessments(): Promise<Application[]> {
+    return this.applicationRepo
+      .createQueryBuilder("application")
+      .select(["application.id", "studentAssessment.id"])
+      .innerJoin("application.studentAssessments", "studentAssessment")
+      .andWhere(
+        "studentAssessment.studentAssessmentStatus = :submittedStatus",
+        {
+          submittedStatus: StudentAssessmentStatus.CancellationRequested,
+        },
+      )
+      .andWhere(`NOT EXISTS (${this.inProgressStatusesExistsQuery})`)
+      .setParameter("inProgressStatuses", [
+        StudentAssessmentStatus.CancellationQueued,
+      ])
```

Comment:
It will be nice to add a comment here to explain why this check is added, as this scenario cannot happen as per our application logic as of today.

```ts 
\\This condition may not be directly related to the current application logic but  is has been added to ensure that at any given time now or future the queue does not pick an application to 
\\cancel when there is already an assessment which is queued for cancellation.

```

Reply:
same for line 113,114

Reply:
Thanks for adding the comment.